### PR TITLE
feat(ng): promote NG formula

### PR DESCRIPTION
**Pull request by sticky-note**
This PR include:
- Promotion of NG formula
- ~~Introduction of recommended `tplroot` var~~
- ~~`{%-` consistency~~

PS: I've not yet tested these changes. 
I began this work to have feedback first.
Ping @myii @n-rodriguez @aboe76

### CODE REVIEW

**Comment by myii** [comment]
@sticky-note In terms of the deprecation messages in cf05b73, if you look at https://github.com/saltstack-formulas/nginx-formula/pull/236/files, you'll see that we have to add these to the states that we're deprecating, i.e. the non-`ng` states that are going to be removed.

**Comment by myii** [comment]
Copying across our earlier conversation from Slack:

> -|-|-
> :-:|---|---
> 00:28 | zabiteus93 | Hi folks, can someone help me with this ? https://github.com/saltstack-formulas/php-formula/pull/183
> 00:33 | myii | @​zabiteus93 Apologies, very busy time at the moment. I saw the PR and was meaning to give you some initial feedback there.
> 00:34 | myii | Now, with `ng` conversions, we've got to do this over stages. We went a little too fast with the `nginx-formula`, with no time to give warnings beforehand.
> 00:35 | myii | Since then, we put the warnings in retrospectively but that really should have been the first stage.
> 00:35 | myii | So taking your PR in general, that's a bit further down the line -- we have to build towards it.
> 00:36 | myii | What we need first is a deprecation warning in the next "release" of `php-formula`. It can use a similar method to the one used by `nginx-formula`.
> 00:38 | myii | This PR is the basis for that, which links to the others: https://github.com/saltstack-formulas/nginx-formula/pull/236.
> 00:40 | myii | This time around, since we're only warning at first, change both `test.fail_without_changes` and `failhard: True` -- these will be reintroduced once the deprecation has taken place.
> 00:41 | myii | And the message should be modified likewise.
> 00:41 | myii | All of this requires another PR first. Are you ready to do that @​zabiteus93?
> 02:38 | zabiteus93 | I have to dig into some regression first and yes, I'm really enthusiast to do that
> 02:39 | zabiteus93 | If somenone can help me with some test, it'll be really aspreciated
> 05:11 | zabiteus93 | deprecated state done : //github.com/saltstack-formulas/nginx-formula/pull/236Debian 9 travis check fails, don't know why yet

**Comment by myii** [comment]
@sticky-note As I mentioned in Slack, it is better if we merge the deprecations warnings _first_, in a separate PR.  We want to release a version that starts warning people to get ready to move across to `php:ng`.  That way, we don't have too much going on in this PR as well.  It is already too much to review, with over 200+ files.  It would be great if we could somehow do this in smaller stages.

Any suggestions @n-rodriguez, @aboe76, @daks?

**Comment by johnkeates** [comment]
Same as with nginx:ng (but that went wrong IIRC)

**Comment by daks** [comment]
In fact, nginx NG promotion has been done quite quickly and we have learnt from it, so I'm also in favour of doing a release to add the deprecation warnings.

**Comment by n-rodriguez** [comment]
> I'm also in favour of doing a release to add the deprecation warnings.

:+1:

**Comment by sticky-note** [comment]
Just corrected
```
nginx.deprecated

**Comment by sticky-note** [comment]
> @sticky-note In terms of the deprecation messages in [cf05b73](https://github.com/saltstack-formulas/php-formula/commit/cf05b73e86d848eb426b880a4a358b9cc67733b0), if you look at https://github.com/saltstack-formulas/nginx-formula/pull/236/files, you'll see that we have to add these to the states that we're deprecating, i.e. the non-`ng` states that are going to be removed.

The non `ng` states present before Promoting are only `fpm.sls` and `cli.sls` that are now folders containing an `init.sls`
What do we have to do with them ?

**Comment by myii** [comment]
> First merge a PR with deprecation warnings with a major release, then Promote `ng` with another major release ?

@sticky-note That's right but the deprecation warnings PR doesn't have to be a major release.  A couple of points about what needs to be done:

1. Each state file under `php` (non-`ng`) needs a deprecation warning included, explaining that these files will be removed in upcoming `v1.0.0`.
1. Each state file under `php.ng` needs an explanatory message included, informing that these files will be promoted to `php` in `v1.0.0` (i.e. that the `.ng` will be dropped).

Obviously, all of these using an `include` at the top of the file.

**Comment by sticky-note** [comment]
@myii I get time to edit this PR
I had to merge master into this branch
How can I get rid of the merge commit ?

**Comment by myii** [comment]
Just merged #167, so a little more rebasing here @sticky-note, sorry!

**Comment by myii** [comment]
@sticky-note Hold on, I'll try to get you something patched up.

**Comment by myii** [comment]
@sticky-note OK, I've patched things up and put it together in one commit on top of the latest upstream here: https://github.com/myii/php-formula/commit/7fbb40b007464c62581dbc4d2a933016be3406a3.  Have a look and see how that compares with the work you have done.

The problem here is that too many changes are being made at the same time, which makes it difficult for `git` to follow what's going on.  For example, when patching this again, it kept tripping up on:

* All of the `{%` which have been fixed to `{%-`
* The introduction of `tplroot` via. `{%- set tplroot = tpldir.split('/')[0] %}`

While these are useful changes (to make eventually), these should be left for a subsequent PR, so that this one can be applied cleanly first.  In fact, it may still be the sensible option to do that, so that we can be completely confident that these changes are sound.  We can always compare with the [linked branch](https://github.com/myii/php-formula/commit/7fbb40b007464c62581dbc4d2a933016be3406a3), to reintroduce the extra changes in a subsequent PR.

**Comment by sticky-note** [comment]
I've removed `php.ng.deprecated.sls` and`php.deprecated.sls` from `pre-commit_semantic-release.sh` and `release.config`

**Comment by myii** [comment]
@sticky-note This is what we did in the the [`nginx-formula`](https://github.com/saltstack-formulas/nginx-formula/tree/c00c058076a9b220ca866c43ba7427d9dcf1ed37/nginx/ng).

**Comment by sticky-note** [comment]
@myii And voilà ?

**Comment by myii** [comment]
@sticky-note Fantastic effort!  I haven't looked in specific detail (the very latest changes) but this really looks like it's made a lot of progress.  We really need some testing on this as well.  I don't use this formula for real, so let's try to ping some of the main contributors to see if they can review/test this PR before we perform this deprecation.

CC: @johnkeates @daks @n-rodriguez @alxwr @aboe76 @arthurlogilab.

**Comment by myii** [comment]
@n-rodriguez The `debian` test fails intermittently because the `php` package sometimes isn't installed.  Should the test be updated to check for the `php7.0` package instead?

**Comment by aboe76** [comment]
will try to to test this, on the weekends, hopefully I can find some time.

**Comment by aboe76** [comment]
@sticky-note first test missing ```Arch``` support in map.jinja.

**Comment by aboe76** [comment]
@sticky-note on ```Debian``` I got a nice deprecated warning message and I see no issues

**Comment by myii** [comment]
@aboe76 @sticky-note Just be aware that this PR will need to include the fix in #189.

**Comment by myii** [comment]
Or probably better to rebase on it once it is merged.

**Comment by myii** [comment]
@sticky-note I've merged #189 and a new release has been created.  I've performed the rebase of this PR locally, including fixing the commit message to ensure that a new release is created.  Do you want me to push it here instead of this commit (obviously still with your authoring details)?

**Comment by sticky-note** [comment]
I've managed to rebase from actual master
I've fixed the release number on `README.rst`
I don(t know debian 9 test is not passing any more

File: docs/README.rst [hunk 1]
```diff
@@ -21,25 +21,28 @@ Formula to set up and configure php
...
+       v1.0.0 <https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0>`_,
+       this formula provided two methods for managing PHP;
+       the old method under ``php`` and the new method under ``php.ng``.
+       The old method has now been removed and ``php.ng`` has been promoted to
+       be ``php`` in its place.
+
+       If you are not in a position to migrate, please pin your repo to the final
+       release tag before
        `v1.0.0 <https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0>`_,
        which is expected to be
```
Comment (myii) [inline docs/README.rst]:
```suggestion
        i.e.
```

File: php/deprecated.sls [hunk 2]
```diff
@@ -1,41 +1,31 @@
...
 
-{#- Evaluating as `False` by default, using this method since `defaults.yaml` not available in this repo #}
-{%- if not salt['config.get']('php').get('warning_messages', {}).get('v1.0.0', {}).get('mute_critical', False) %}
-php-deprecated-in-v1.0.0-test-succeed:
-  test.succeed_without_changes:
+php-deprecated-in-v1.0.0-test-fail:
+  test.fail_without_changes:
     - name: |
 
-
```
Comment (myii) [inline php/deprecated.sls]: Don't remove this line -- we need two blank lines here to match the `deprecated.sls` from `nginx-formula` ([link](https://github.com/saltstack-formulas/nginx-formula/blob/df649c997bc2d4bea7721fbbb54c23efa4688f0e/nginx/deprecated.sls)).  This is best for displaying this message.

**Comment by myii** [comment]
@sticky-note As for the Debian failure, that's what I mentioned above: https://github.com/saltstack-formulas/php-formula/pull/183#issuecomment-519291286.  See:

* https://travis-ci.com/saltstack-formulas/php-formula/jobs/224627690#L712
* https://travis-ci.com/saltstack-formulas/php-formula/jobs/224627690#L768

We can solve this in a tiny separate PR before/after this PR.

**Comment by sticky-note** [comment]
I confirm `php7.0` only is available from official Debian repos

**Comment by sticky-note** [comment]
Made the two requested changes
And now tests pass in Ubuntu and Debian with that PR
https://github.com/saltstack-formulas/php-formula/pull/190
Thanks @myii

**Comment by myii** [comment]
@sticky-note Thanks for the changes.  Let's wait a little bit, so that #190 can be merged and that any other contributors can check this PR.

@aboe76 How many reviews do you think we need before merging this?

**Comment by aboe76** [comment]
If we have a confirmation that it works on the major distro's I'm fine:
- Debian/Ubuntu
- Redhat/Centos

The others can be fixed with later PR's,

**Comment by sticky-note** [comment]
I've just rebased from master and fixed release number on `README.rst` and `php/deprecated.sls`.
Can I remove the WIP prefix on the PR name ?

**Comment by myii** [comment]
@sticky-note Yes, please do that. This is almost ready for merging now, thanks.

File: php/deprecated.sls [hunk 3]
```diff
@@ -1,41 +1,32 @@
...
         #                                                                              #
-        # This formula currently provides two methods for managing PHP; the old method #
-        # under `php` and the new method under `php.ng`. In upcoming `v1.0.0`, the old #
-        # method will be removed and `php.ng` will be promoted to `php` in its place.  #
+        # Prior to `v1.0.0`, this formula provided two methods for managing PHP; the   #
+        # old method under `php` and the new method under `php.ng`. The old method     #
+        # has now been removed and `php.ng` has been promoted to be `php` in its       #
+        # place.                                                                       #
         #                                                                              #
         # If you are not in a position to migrate, you will need to pin your repo to   #
```
Comment (myii) [inline php/deprecated.sls]:
```suggestion
         # If you are not in a position to migrate, please pin your repo to the final   #
```

File: php/deprecated.sls [hunk 4]
```diff
@@ -1,41 +1,32 @@
...
         # the final release tag before `v1.0.0`, which is expected to be `v0.39.2`.    #
```
Comment (myii) [inline php/deprecated.sls]:
```suggestion
         # release tag before `v1.0.0`, i.e. `v0.39.2`.                                 #
```

File: docs/README.rst [hunk 5]
```diff
@@ -21,25 +21,28 @@ Formula to set up and configure php
...
-   * - This formula currently provides two methods for managing PHP; the old method
-       under ``php`` and the new method under ``php.ng``.
-       In upcoming `v1.0.0 <https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0>`_,
-       the old method will be removed and ``php.ng`` will be promoted to ``php`` in its place.
-
-       If you are not in a position to migrate, you will need to pin your repo to
-       the final release tag before
+   * - WARNING: BREAKING CHANGES SINCE ``v1.0.0``
+   * - Prior to
+       v1.0.0 <https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0>`_,
```
Comment (myii) [inline docs/README.rst]:
```suggestion
       `v1.0.0 <https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0>`_,
```

**Comment by myii** [comment]
> If we have a confirmation that it works on the major distro's I'm fine:
> 
> * Debian/Ubuntu
> * Redhat/Centos

@aboe76 You mentioned that you tested Debian above, so just the other three remain?

**Comment by aboe76** [comment]
@myii you are correct but having others testing it with different parameters would provide a better result.

**Comment by sticky-note** [comment]
Okay for minor edits.
Works for me on `debian 9` and `FreeBSD 12`

**Comment by aboe76** [comment]
Can somebody confirm this works on Centos/RedHat or Fedora?

**Comment by myii** [comment]
@philpep (and @arthurlogilab) may be able to help us with CentOS: https://github.com/saltstack-formulas/php-formula/pull/191#issuecomment-521274025.

**Comment by myii** [comment]
@sticky-note I'm sorry, I was holding back on #191 so that this could be merged first, in order to avoid _another_ rebase, but I didn't put a warning on my own PR #192!  Since that was merged and this needs to be rebased anyway, I've merged #191 as well.  I'm happy to perform the rebase on this, since it's my fault in a way.  Are you OK with that?

**Comment by myii** [comment]
@sticky-note So I've performed the rebase on my fork, including incorporating all of the PRs that have been merged since this PR was started:

* https://github.com/myii/php-formula/commit/5aade39da876cab70676f2b7dbfd89e56e650599

I hope this is the last time we have to do this because it takes a lot of effort.

Either I can push the commit here directly, or you can at least compare to that commit when performing the rebase yourself.  As you can see, the commit is still authored with your details.  It's up to you how you want to proceed.

**Comment by sticky-note** [comment]
@myii, I've rebased with `master`
Not sure if `pillar.example` is correct

File: php/map.jinja [hunk 6]
```diff
@@ -1,500 +1,3247 @@
...
+                            'suhosin5_ext': 'suhosin.so',
+                            'suhosin5_repo': 'https://github.com/sektioneins/suhosin',
+                            'suhosin7_ext': 'suhosin7.so',
+                            'suhosin7_repo': 'https://github.com/sektioneins/suhosin7',
+                            'sybase': 'php' + php_version + '-sybase',
+                            'tcpdf': 'php-tcpdf',
+                            'temp_dir': '/tmp',
+                            'tidy': 'php' + php_version + '-tidy',
+                            'xdebug': 'php' + php_version + '-xdebug',
+                            'xml': 'php' + php_version + '-xml',
```
Comment (myii) [inline php/map.jinja]:
```suggestion
                            'xml': ['php' + php_version + '-xml', 'php' + php_version + '-xmlrpc'],
```

File: php/map.jinja [hunk 7]
```diff
@@ -1,500 +1,3247 @@
...
```
Comment (myii) [inline php/map.jinja]:
```suggestion
                            'xml': ['php' + php_version + '-xml', 'php' + php_version + '-xmlrpc'],
```

File: php/map.jinja [hunk 8]
```diff
@@ -1,500 +1,3247 @@
...
+                        'settings': {},
+                    }
+                },
+                'ini': {
+                    'defaults': {
+                        'PHP': {
+                            'allow_url_fopen': 'On',
+                            'allow_url_include': 'Off',
+                            'asp_tags': 'Off',
+                            'auto_globals_jit': 'On
```
Comment (myii) [inline php/map.jinja]:
```suggestion
                            'xml': ['php' + php_version + '-xml', 'php' + php_version + '-xmlrpc'],
```

Comment (myii) [reply php/map.jinja]: @sticky-note GitHub seems to be struggling here -- I've clearly indicated line 2267 but it's showing line 1266!  Hopefully, it's clear to you what I'm indicating.

**Comment by myii** [comment]
> I've rebased with `master`

Sorry again for the trouble.  Let's hope this is the last time.

> Not sure if `pillar.example` is correct

It appears to be fine.  What do you think is incorrect?

File: php/map.jinja (hunk 6 above)
Comment (myii) [reply php/map.jinja]: All of these changes are from the merge of #175, which took place after this PR was first started.

**Comment by sticky-note** [comment]
@myii 
> It appears to be fine. What do you think is incorrect?
I've not digged in details to the very last modifications.
I've just trusted the mergetool with some fixes concerning `ng`
If you think it is correct, so go to finalize this PR.
What to to next ?

**Comment by myii** [comment]
> What to to next ?

@sticky-note That's all from my end and hopefully from your end.  We're just waiting for the final confirmations before this can be merged.  Looking above, it's what @aboe76 has mentioned:

> Can somebody confirm this works on Centos/RedHat or Fedora?

@philpep @arthurlogilab Can either of you do that?

**Comment by sticky-note** [comment]
Want to test it on `CentOS` with `vagrant` to save some  time. Is it ok @myii ?

**Comment by myii** [comment]
> Want to test it on `CentOS` with `vagrant` to save some time. Is it ok @myii ?

Sure, it just needs to be confirmed and then we should be able to proceed.  Is that OK, @aboe76?

**Comment by sticky-note** [comment]
@myii @aboe76 States for `php-fpm`, `fpm-pools` and around twenty `php` modules, and `composer` installation with `scl repo` enabled and `php_version : 72` are confirmed to work on a `CentOS 7` vagrant box

**Comment by myii** [comment]
Thanks @sticky-note, I'm happy for this to be merged.  Any final comments @aboe76, @n-rodriguez @daks @johnkeates?

**Comment by aboe76** [comment]
Lgtm

**Comment by n-rodriguez** [comment]
> Lgtm

same here :+1:

**Comment by myii** [comment]
Congratulations @sticky-note, it's merged -- thanks for your hard work and patience!

Appreciate all of the the feedback and reviews from everyone else.

**Comment by myii** [comment]
@sticky-note The commit message fixes must have got lost during all of the rebases, so it didn't trigger a `v1.0.0`.  See https://github.com/myii/php-formula/commit/5aade39da876cab70676f2b7dbfd89e56e650599:

```
BREAKING CHANGE: all previous `php` based configurations must be reviewed;
`php.ng` usage must be promoted to `php` and any uses of the original
`php` will have to be converted.
```

Since `BREAKING CHANGES` (extra `S`) was used instead, that's why this happened.  Don't worry, I'll add a commit now to fix this.

**Comment by myii** [comment]
Resolved in 57b37dd.  Welcome to [`v1.0.0`](https://github.com/saltstack-formulas/php-formula/releases/tag/v1.0.0)!

**Pull request merged**